### PR TITLE
[baremetal] micro-optimisations for Thumb and small configs

### PR DESCRIPTION
**Pull request by kjbracey-arm**
Various changes to reduce code size - saves nearly 400 bytes in a minimal client-only DTLS config on Cortex-M0+.

* Fix `MBEDTLS_SSL_CONF_ENDPOINT` flagging - didn't compile when activated.
* Add `MBEDTLS_SSL_CONF_TRANSPORT` following the endpoint model.
* Avoid bitfields.
* Reorder structures.

More detail in individual commit messages.

### CODE REVIEW

**Comment by yogpan01** [comment]
@shelib01 @AndrzejKurek Please review.

File: include/mbedtls/ssl_internal.h [hunk 1]
```diff
@@ -1454,6 +1454,21 @@ static inline unsigned int mbedtls_ssl_conf_get_endpoint(
 }
 #endif /* MBEDTLS_SSL_CONF_ENDPOINT */
 
+#if !defined(MBEDTLS_SSL_CONF_TRANSPORT)
+static inline unsigned int mbedtls_ssl_conf_get_transport(
+    mbedtls_ssl_config  const *conf )
```
Comment (AndrzejKurek) [inline include/mbedtls/ssl_internal.h:1459]:
```suggestion
    mbedtls_ssl_config const *conf )
```

File: include/mbedtls/ssl_internal.h [hunk 2]
```diff
@@ -1454,6 +1454,21 @@ static inline unsigned int mbedtls_ssl_conf_get_endpoint(
...
+{
+    return( conf->transport );
+}
+#else /* !MBEDTLS_SSL_CONF_TRANSPORT */
+static inline unsigned int mbedtls_ssl_conf_get_transport(
+    mbedtls_ssl_config  const *conf )
```
Comment (AndrzejKurek) [inline include/mbedtls/ssl_internal.h:1465]:
```suggestion
    mbedtls_ssl_config const *conf )
```

**Review by AndrzejKurek — CHANGES_REQUESTED** [review]
It would be great to have `MBEDTLS_SSL_CONF_TRANSPORT` tested with the baremetal config - please add it [here](https://github.com/ARMmbed/mbedtls/blob/baremetal/configs/baremetal.h),

**Comment by AndrzejKurek** [comment]
Rerunning tests due to some seemingly spurious errors.

File: include/mbedtls/ssl_internal.h (hunk 1 above)
Comment (kjbracey-arm) [reply include/mbedtls/ssl_internal.h:1459]: That's been copy-pasted to 20+ places in the file. Shall I fix them all, or just have this one correct+inconsistent?

Comment (AndrzejKurek) [reply include/mbedtls/ssl_internal.h:1459]: Good point, please just fix your code. I'll create a PR for the rest.

**Comment by kjbracey-arm** [comment]
> It would be great to have `MBEDTLS_SSL_CONF_TRANSPORT` tested with the baremetal config - please add it [here](https://github.com/ARMmbed/mbedtls/blob/baremetal/configs/baremetal.h).

I've also added the `MBEDTLS_SSL_CONF_ENDPOINT` that would have shown the compile error.

**Comment by AndrzejKurek** [comment]
> > It would be great to have `MBEDTLS_SSL_CONF_TRANSPORT` tested with the baremetal config - please add it [here](https://github.com/ARMmbed/mbedtls/blob/baremetal/configs/baremetal.h).
> 
> I've also added the `MBEDTLS_SSL_CONF_ENDPOINT` that would have shown its compile error.

This change might be problematic, as both server and client in tests use the same baremetal config. MFL tests might fail - we'll see.

**Comment by kjbracey-arm** [comment]
> This change might be problematic, as both server and client in tests use the same baremetal config. MFL tests might fail - we'll see.

Ah, I thought it was viable because I saw `CLI_C` without `SRV_C`, but the `SRV_C` was elsewhere.

**Comment by kjbracey-arm** [comment]
Dropped the `CONF_ENDPOINT` in the latest update round.

**Comment by kjbracey-arm** [comment]
How are the tests looking? There's an ABI change (expected), and then all the Travis failures that always seem to happen, and not sure if there's anything else I need to look at.

**Review by AndrzejKurek — APPROVED** [review]
Looks good.

**Review by shelib01 — APPROVED** [review]

**Pull request merged**